market/retrieval/remoteblockstore/remoteblockstore.go

```diff
@@ -0,0 +1,287 @@
+		GetSuccessResponseCount:     HttpRblsGetSuccessResponseCount,
+		BytesSentCount:              HttpRblsBytesSentCount,
+		HasRequestCount:             HttpRblsHasRequestCount,
+		HasFailResponseCount:        HttpRblsHasFailResponseCount,
+		HasSuccessResponseCount:     HttpRblsHasSuccessResponseCount,
+		GetSizeRequestCount:         HttpRblsGetSizeRequestCount,
+		GetSizeFailResponseCount:    HttpRblsGetSizeFailResponseCount,
+		GetSizeSuccessResponseCount: HttpRblsGetSizeSuccessResponseCount,
+	}
+
+	ro := &RemoteBlockstore{
+		idxApi:           api,
+		blockMetrics:     httpBlockMetrics,
+		db:               db,
+		pp:               pp,
+		pieceReaderCache: prCache,
+	}
+
+	expireCallback := func(key string, reason ttlcache.EvictionReason, value interface{}) {
+		log.Debugw("expire callback", "piececid", key, "reason", reason)
+
+		r := value.(*cachedSectionReader)
+
+		ro.pieceReaderCacheMu.Lock()
+		defer ro.pieceReaderCacheMu.Unlock()
+
+		r.expired = true
+
+		if r.refs <= 0 {
+			r.cancel()
+			return
+		}
+
+		log.Debugw("expire callback with refs > 0", "refs", r.refs, "piececid", key, "reason", reason)
+	}
+
+	prCache.SetExpirationReasonCallback(expireCallback)
+
+	return ro
+}
+
+func (ro *RemoteBlockstore) Get(ctx context.Context, c cid.Cid) (b blocks.Block, err error) {
+	if ro.blockMetrics != nil {
+		stats.Record(ctx, ro.blockMetrics.GetRequestCount.M(1))
+	}
+
+	log.Debugw("Get", "cid", c)
+
+	// Get the pieces that contain the cid
+	pieces, err := ro.idxApi.PiecesContainingMultihash(ctx, c.Hash())
+
+	// Check if it's an identity cid, if it is, return its digest
+	if err != nil {
+		digest, ok, iderr := isIdentity(c)
+		if iderr == nil && ok {
+			if ro.blockMetrics != nil {
+				stats.Record(ctx, ro.blockMetrics.GetSuccessResponseCount.M(1))
+			}
+			return blocks.NewBlockWithCid(digest, c)
+		}
+		if ro.blockMetrics != nil {
+			stats.Record(ctx, ro.blockMetrics.GetFailResponseCount.M(1))
+		}
+		return nil, fmt.Errorf("getting pieces containing cid %s: %w", c, err)
+	}
+	if len(pieces) == 0 {
+		return nil, fmt.Errorf("no pieces with cid %s found", c)
+	}
+
+	// Get a reader over one of the pieces and extract the block data
+	var merr error
+	for _, pieceCid := range pieces {
+		data, err := func() ([]byte, error) {
+			// Get a reader over the piece data
+			reader, err := ro.GetSharedPieceReader(ctx, pieceCid)
+			if err != nil {
+				return nil, fmt.Errorf("getting piece reader: %w", err)
+			}
+			defer func(reader SectionReader) {
+				_ = reader.Close()
+			}(reader)
+
+			// Get the offset of the block within the piece (CAR file)
+			offsetSize, err := ro.idxApi.GetOffsetSize(ctx, pieceCid, c.Hash())
+			if err != nil {
+				return nil, fmt.Errorf("getting offset/size for cid %s in piece %s: %w", c, pieceCid, err)
+			}
+
+			// Seek to the section offset
+			readerAt := readerutil.NewReadSeekerFromReaderAt(reader, int64(offsetSize.Offset))
+			// Read the block data
+			bufferSize := 4096
```

Comment:
Why do we not just read the correct size here?

Reply:
That was a work around for 2k sectors in Boost. Now that we can reproduce the pieceReader panic, I have removed the custom buffer size and went with default.